src/Robot.Drivers/Sonar/HcSr04DistanceSensorDriver.cs

```diff
@@ -0,0 +1,79 @@
+﻿using Iot.Device.Hcsr04;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Device.Gpio;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Robot.Drivers.Sonar
+{
+    public class HcSr04DistanceSensorDriver : IDistanceSensor
+    {
+        private Hcsr04 Sensor;
+
+        public int Id { get; }
+        public double Angle { get; }
+        public CancellationTokenSource MeasuringCancellationToken { get; private set; }
+        public HcSr04DistanceSensorDriverSettings Settings { get; }
+        public ILogger<HcSr04DistanceSensorDriver> Logger { get; }
+
+        public event EventHandler<SonarDistanceEventArgs> SonarDistanceChanged;
+
+        public HcSr04DistanceSensorDriver(HcSr04DistanceSensorDriverSettings settings, ILogger<HcSr04DistanceSensorDriver> logger)
+        {
+            Id = settings.SensorId;
+            Angle = settings.Angle;
+            Sensor = new Hcsr04(settings.TriggerPin, settings.EchoPin, PinNumberingScheme.Logical);
+            MeasuringCancellationToken = new CancellationTokenSource();
+            Task.Run(MeasurementCycle);
+            Settings = settings;
+            Logger = logger;
+        }
+
+        private async void MeasurementCycle()
```

Comment:
I think we will have issues with this. Sonar sensors need to be fired sequentially and not overlap. Otherwise we will read the response that was intended for another sensor and get fake readings.

Reply:
Thanks, that is a very reasonable concern, I missed that.
I took a look at the implementation ([here's the link](https://github.com/dotnet/iot/blob/master/src/devices/Hcsr04/Hcsr04.cs))
It looks like there will be no issue with it unless our sensors share one echo pin.
So could you also take a look at the sensor code and confirm that the concern remains?

Reply:
Issue remains, but I'd like to get into the code base anyway.
The problem are not the pins but the sound wave in the air. We'll talk in the week, its not a big deal for now